Type Error when implementing Vue "Plugin" type in TypeScript

### Vue version

3.3.4

### Link to minimal reproduction

https://stackblitz.com/edit/vitejs-vite-gbgbxq?terminal=build

### Steps to reproduce

[Stackbliz link](https://stackblitz.com/edit/vitejs-vite-gbgbxq?terminal=build) shows the TS compile error in plugin.ts.
Run `npm run build`

This example shows a type error in vue plugin implementation.

`src/plugins/plugin.ts:9:41 - error TS2422: A class can only implement an object type or intersection of object types with statically known members.`

### What is expected?

Vue types provide **Plugin**, **PluginInstallFunction** and others for plugin implementation. It should be properly used for _implementation_ of plugins without Type Error as above.
I would like to ask you to consider revising Type definitions for Plugin.


### What is actually happening?

This vue plugin is a class implementation which has 'install' function, which comply with a general guidance of vue plugin implementation.

In a sense of strict type manner for vue plugin implementation, this example tried to "_implements_" vue **Plugin** type to the class. But, the **Plugin** type is not working with class's "_implements_" directive of TypeScript, since the type is likely dynamic and undetermined for the target class or so on. Thus, we see the following error when compiling `*.ts`.

`src/plugins/plugin.ts:9:41 - error TS2422: A class can only implement an object type or intersection of object types with statically known members.`

### System Info

```shell
System:
    OS: Linux 5.15 Ubuntu 22.04.2 LTS 22.04.2 LTS (Jammy Jellyfish)
    CPU: (2) x64 Intel(R) Core(TM)2 Duo CPU     P9700  @ 2.80GHz
    Memory: 1.79 GB / 7.70 GB
    Container: Yes
    Shell: 5.1.16 - /bin/bash
  Binaries:
    Node: 19.6.0 - ~/.nvm/versions/node/v19.6.0/bin/node
    Yarn: 1.22.19 - /usr/bin/yarn
    npm: 9.4.0 - ~/.nvm/versions/node/v19.6.0/bin/npm
```


### Any additional comments?

Thanks in advance for your taking a look for this issue!

--- a/packages/runtime-core/src/apiCreateApp.ts
+++ b/packages/runtime-core/src/apiCreateApp.ts
@@ -155,17 +155,19 @@ export interface AppContext {
   filters?: Record<string, Function>
 }
 
-type PluginInstallFunction<Options> = Options extends unknown[]
+type PluginInstallFunction<Options = any[]> = Options extends unknown[]
   ? (app: App, ...options: Options) => any
   : (app: App, options: Options) => any
 
+export type ObjectPlugin<Options = any[]> = {
+  install: PluginInstallFunction<Options>
+}
+export type FunctionPlugin<Options = any[]> = PluginInstallFunction<Options> &
+  Partial<ObjectPlugin<Options>>
+
 export type Plugin<Options = any[]> =
-  | (PluginInstallFunction<Options> & {
-      install?: PluginInstallFunction<Options>
-    })
-  | {
-      install: PluginInstallFunction<Options>
-    }
+  | FunctionPlugin<Options>
+  | ObjectPlugin<Options>
 
 export function createAppContext(): AppContext {
   return {

--- a/packages/runtime-core/src/index.ts
+++ b/packages/runtime-core/src/index.ts
@@ -195,9 +195,9 @@ export type {
   App,
   AppConfig,
   AppContext,
+  Plugin,
   ObjectPlugin,
   FunctionPlugin,
-  Plugin,
   CreateAppFunction,
   OptionMergeFunction
 } from './apiCreateApp'

--- a/packages/runtime-core/src/index.ts
+++ b/packages/runtime-core/src/index.ts
@@ -195,6 +195,8 @@ export type {
   App,
   AppConfig,
   AppContext,
+  ObjectPlugin,
+  FunctionPlugin,
   Plugin,
   CreateAppFunction,
   OptionMergeFunction

